src/shell/braces.zig

```diff
@@ -0,0 +1,805 @@
+            .comma => SmolStr.fromChar(','),
+            .text => |txt| txt,
+            .close => SmolStr.fromChar('}'),
+            .eof => SmolStr.empty(),
+        };
+    }
+};
+
+pub const AST = struct {
+    pub const Atom = union(enum) {
+        text: SmolStr,
+        expansion: Expansion,
+    };
+
+    const Group = struct {
+        bubble_up: ?*Group = null,
+        bubble_up_next: ?u16 = null,
+        atoms: union(enum) { single: Atom, many: []Atom },
+    };
+
+    const Expansion = struct {
+        variants: []AST.Group,
+    };
+};
+
+const MAX_NESTED_BRACES = 10;
+
+/// A stack on the stack
+pub fn StackStack(comptime T: type, comptime SizeType: type, comptime N: SizeType) type {
+    return struct {
+        items: [N]T = undefined,
+        len: SizeType = 0,
+
+        pub const Error = error{
+            StackEmpty,
+            StackFull,
+        };
+
+        pub fn top(this: *@This()) ?T {
+            if (this.len == 0) return null;
+            return this.items[this.len - 1];
+        }
+
+        pub fn topPtr(this: *@This()) ?*T {
+            if (this.len == 0) return null;
+            return &this.items[this.len - 1];
+        }
+
+        pub fn push(this: *@This(), value: T) Error!void {
+            if (this.len == N) return Error.StackFull;
+            this.items[this.len] = value;
+            this.len += 1;
+        }
+
+        pub fn pop(this: *@This()) ?T {
+            if (this.top()) |v| {
+                this.len -= 1;
+                return v;
+            }
+            return null;
+        }
+    };
+}
+
+/// This may have false positives but it is fast
+pub fn fastDetect(src: []const u8) bool {
+    const Quote = enum { single, double };
+    _ = Quote;
+
+    var has_open = false;
+    var has_close = false;
+    if (src.len < 16) {
+        for (src) |char| {
+            switch (char) {
+                '{' => {
+                    has_open = true;
+                },
+                '}' => {
+                    has_close = true;
+                },
+            }
+            if (has_close and has_close) return true;
+        }
+        return false;
+    }
+
+    const needles = comptime [2]@Vector(16, u8){
+        @splat('{'),
+        @splat('}'),
+        @splat('"'),
+    };
+
+    const i: usize = 0;
+    while (i + 16 <= src.len) {
+        const haystack = src[i .. i + 16].*;
+        if (std.simd.firstTrue(needles[0] == haystack)) {
+            has_open = true;
+        }
+        if (std.simd.firstTrue(needles[1] == haystack)) {
+            has_close = true;
+        }
+        if (has_open and has_close) return true;
+    }
+
+    if (i < src.len) {
+        for (src) |char| {
+            switch (char) {
+                '{' => {
+                    has_open = true;
+                },
+                '}' => {
+                    has_close = true;
+                },
+            }
+            if (has_close and has_close) return true;
```

Comment:
`has_close and has_close`? is this supposed to be has_close and has_open?

or, what if this was just in the switch itself (twice, one per) checking for the other boolean?